// @coinbase-sample/prime-sdk-ts/dist/client-modular.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCredentialsFromEnv = exports.CoinbasePrimeCredentials = exports.CoinbaseError = exports.CoinbaseClientException = exports.Method = exports.CoinbaseClient = exports.CoinbasePrimeClientWithServices = void 0;
// Modular client entry point - for users who want optimized bundle sizes
// with lazy-loaded services and modular type imports
var clients_1 = require("./clients");
Object.defineProperty(exports, "CoinbasePrimeClientWithServices", { enumerable: true, get: function () { return clients_1.CoinbasePrimeClientWithServices; } });
Object.defineProperty(exports, "CoinbaseClient", { enumerable: true, get: function () { return clients_1.CoinbaseClient; } });
Object.defineProperty(exports, "Method", { enumerable: true, get: function () { return clients_1.Method; } });
Object.defineProperty(exports, "CoinbaseClientException", { enumerable: true, get: function () { return clients_1.CoinbaseClientException; } });
Object.defineProperty(exports, "CoinbaseError", { enumerable: true, get: function () { return clients_1.CoinbaseError; } });
var credentials_1 = require("./credentials");
Object.defineProperty(exports, "CoinbasePrimeCredentials", { enumerable: true, get: function () { return credentials_1.CoinbasePrimeCredentials; } });
var envUtils_1 = require("./shared/envUtils");
Object.defineProperty(exports, "createCredentialsFromEnv", { enumerable: true, get: function () { return envUtils_1.createCredentialsFromEnv; } });
__exportStar(require("./model/enums/"), exports);
// NOTE: Service-specific request/response types are NOT exported here.
// For optimal tree-shaking, import them from their dedicated modules:
//
// import type { CreateOrderRequest } from '@coinbase-sample/prime-sdk-ts/orders';
// import type { ListWalletsRequest } from '@coinbase-sample/prime-sdk-ts/wallets';
// import type { GetActivityRequest } from '@coinbase-sample/prime-sdk-ts/activities';
//
// Or import all service types at once:
// import type * from '@coinbase-sample/prime-sdk-ts/types';
//
// This design ensures the modular client bundle stays minimal while
// still providing access to all service interfaces for TypeScript support.
