// @coinbase-sample/prime-sdk-ts/dist/clients/clientWithServices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinbasePrimeClientWithServices = void 0;
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const core_ts_1 = require("@coinbase-sample/core-ts");
const constants_1 = require("../constants");
const toCamelCase_1 = require("../shared/toCamelCase");
const envUtils_1 = require("../shared/envUtils");
/**
 * Enhanced Coinbase Prime Client with lazy-loaded service getters.
 * Services are only instantiated when first accessed, improving tree-shaking
 * and reducing initial bundle size.
 */
class CoinbasePrimeClientWithServices extends core_ts_1.CoinbaseClient {
    constructor(credentials, apiBasePath, options) {
        const defaultClientOptions = Object.assign({ defaultLimit: constants_1.DEFAULT_PAGINATION_LIMIT, maxPages: constants_1.DEFAULT_MAX_PAGES, maxItems: constants_1.DEFAULT_MAX_ITEMS }, options);
        let basePath = constants_1.API_BASE_PATH;
        if (apiBasePath && apiBasePath.length > 0) {
            basePath = apiBasePath;
        }
        super(basePath, credentials, constants_1.USER_AGENT, defaultClientOptions);
        // transform the response data to camelCase
        this.addTransformResponse((response) => {
            return Object.assign(Object.assign({}, response), { data: (0, toCamelCase_1.toCamelCase)(response.data) });
        });
    }
    /**
     * Create a client from environment variables
     * Requires PRIME_CREDENTIALS environment variable with JSON containing:
     * { "AccessKey": "...", "SecretKey": "...", "Passphrase": "..." }
     */
    static fromEnv(baseUrl, options) {
        const credentials = (0, envUtils_1.createCredentialsFromEnv)();
        return new CoinbasePrimeClientWithServices(credentials, baseUrl, options);
    }
    /**
     * Lazy getter for ActivitiesService
     * @example
     * ```typescript
     * const activities = await client.activities.listPortfolioActivities({ portfolioId });
     * ```
     */
    get activities() {
        if (!this._activitiesService) {
            const { ActivitiesService } = require('../activities');
            this._activitiesService = new ActivitiesService(this);
        }
        return this._activitiesService;
    }
    /**
     * Lazy getter for AddressBooksService
     * @example
     * ```typescript
     * const addressBooks = await client.addressBooks.listAddressBooks({ portfolioId });
     * ```
     */
    get addressBooks() {
        if (!this._addressBooksService) {
            const { AddressBooksService } = require('../addressBooks');
            this._addressBooksService = new AddressBooksService(this);
        }
        return this._addressBooksService;
    }
    /**
     * Lazy getter for AllocationService
     * @example
     * ```typescript
     * const allocation = await client.allocations.createAllocation({ portfolioId, ... });
     * ```
     */
    get allocations() {
        if (!this._allocationService) {
            const { AllocationService } = require('../allocations');
            this._allocationService = new AllocationService(this);
        }
        return this._allocationService;
    }
    /**
     * Lazy getter for AssetsService
     * @example
     * ```typescript
     * const assets = await client.assets.listAssets({ entityId });
     * ```
     */
    get assets() {
        if (!this._assetsService) {
            const { AssetsService } = require('../assets');
            this._assetsService = new AssetsService(this);
        }
        return this._assetsService;
    }
    /**
     * Lazy getter for BalancesService
     * @example
     * ```typescript
     * const balance = await client.balances.getWalletBalance({ portfolioId, walletId });
     * ```
     */
    get balances() {
        if (!this._balancesService) {
            const { BalancesService } = require('../balances');
            this._balancesService = new BalancesService(this);
        }
        return this._balancesService;
    }
    /**
     * Lazy getter for CommissionService
     * @example
     * ```typescript
     * const commission = await client.commission.getPortfolioCommission({ portfolioId });
     * ```
     */
    get commission() {
        if (!this._commissionService) {
            const { CommissionService } = require('../commission');
            this._commissionService = new CommissionService(this);
        }
        return this._commissionService;
    }
    /**
     * Lazy getter for FinancingService
     * @example
     * ```typescript
     * const buyingPower = await client.financing.getPortfolioBuyingPower({ portfolioId });
     * ```
     */
    get financing() {
        if (!this._financingService) {
            const { FinancingService } = require('../financing');
            this._financingService = new FinancingService(this);
        }
        return this._financingService;
    }
    /**
     * Lazy getter for FuturesService
     * @example
     * ```typescript
     * const positions = await client.futures.getEntityFuturesPositions({ entityId });
     * ```
     */
    get futures() {
        if (!this._futuresService) {
            const { FuturesService } = require('../futures');
            this._futuresService = new FuturesService(this);
        }
        return this._futuresService;
    }
    /**
     * Lazy getter for InvoicesService
     * @example
     * ```typescript
     * const invoices = await client.invoices.listInvoices({ entityId });
     * ```
     */
    get invoices() {
        if (!this._invoicesService) {
            const { InvoicesService } = require('../invoices');
            this._invoicesService = new InvoicesService(this);
        }
        return this._invoicesService;
    }
    /**
     * Lazy getter for OnchainAddressBookService
     * @example
     * ```typescript
     * const addressBook = await client.onchainAddressBook.listOnchainAddressBook({ portfolioId });
     * ```
     */
    get onchainAddressBook() {
        if (!this._onchainAddressBookService) {
            const { OnchainAddressBookService } = require('../onchainAddressBook');
            this._onchainAddressBookService = new OnchainAddressBookService(this);
        }
        return this._onchainAddressBookService;
    }
    /**
     * Lazy getter for OrdersService
     * @example
     * ```typescript
     * const order = await client.orders.createOrder({
     *   portfolioId: "123",
     *   side: "BUY",
     *   productId: "BTC-USD",
     *   type: "MARKET",
     *   baseQuantity: "0.001"
     * });
     * ```
     */
    get orders() {
        if (!this._ordersService) {
            const { OrdersService } = require('../orders');
            this._ordersService = new OrdersService(this);
        }
        return this._ordersService;
    }
    /**
     * Lazy getter for PaymentMethodsService
     * @example
     * ```typescript
     * const paymentMethods = await client.paymentMethods.listEntityPaymentMethods({ entityId });
     * ```
     */
    get paymentMethods() {
        if (!this._paymentMethodsService) {
            const { PaymentMethodsService } = require('../paymentMethods');
            this._paymentMethodsService = new PaymentMethodsService(this);
        }
        return this._paymentMethodsService;
    }
    /**
     * Lazy getter for PortfoliosService
     * @example
     * ```typescript
     * const portfolios = await client.portfolios.listPortfolios();
     * ```
     */
    get portfolios() {
        if (!this._portfoliosService) {
            const { PortfoliosService } = require('../portfolios');
            this._portfoliosService = new PortfoliosService(this);
        }
        return this._portfoliosService;
    }
    /**
     * Lazy getter for PositionsService
     * @example
     * ```typescript
     * const positions = await client.positions.listAggregateEntityPositions({ entityId });
     * ```
     */
    get positions() {
        if (!this._positionsService) {
            const { PositionsService } = require('../positions');
            this._positionsService = new PositionsService(this);
        }
        return this._positionsService;
    }
    /**
     * Lazy getter for ProductsService
     * @example
     * ```typescript
     * const products = await client.products.listProducts({ portfolioId });
     * ```
     */
    get products() {
        if (!this._productsService) {
            const { ProductsService } = require('../products');
            this._productsService = new ProductsService(this);
        }
        return this._productsService;
    }
    /**
     * Lazy getter for StakingService
     * @example
     * ```typescript
     * const stake = await client.staking.createStake({ walletId, assetId, amount });
     * ```
     */
    get staking() {
        if (!this._stakingService) {
            const { StakingService } = require('../staking');
            this._stakingService = new StakingService(this);
        }
        return this._stakingService;
    }
    /**
     * Lazy getter for TransactionsService
     * @example
     * ```typescript
     * const transaction = await client.transactions.getTransaction({ portfolioId, transactionId });
     * ```
     */
    get transactions() {
        if (!this._transactionsService) {
            const { TransactionsService } = require('../transactions');
            this._transactionsService = new TransactionsService(this);
        }
        return this._transactionsService;
    }
    /**
     * Lazy getter for UsersService
     * @example
     * ```typescript
     * const users = await client.users.listUsers({ entityId });
     * ```
     */
    get users() {
        if (!this._usersService) {
            const { UsersService } = require('../users');
            this._usersService = new UsersService(this);
        }
        return this._usersService;
    }
    /**
     * Lazy getter for WalletsService
     * @example
     * ```typescript
     * const wallets = await client.wallets.listWallets({ portfolioId });
     * ```
     */
    get wallets() {
        if (!this._walletsService) {
            const { WalletsService } = require('../wallets');
            this._walletsService = new WalletsService(this);
        }
        return this._walletsService;
    }
}
exports.CoinbasePrimeClientWithServices = CoinbasePrimeClientWithServices;
