// @coinbase-sample/prime-sdk-ts/dist/clients/index.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinbaseError = exports.CoinbaseClientException = exports.Method = exports.CoinbaseClient = exports.CoinbasePrimeClientWithServices = exports.CoinbasePrimeClient = void 0;
var client_1 = require("./client");
Object.defineProperty(exports, "CoinbasePrimeClient", { enumerable: true, get: function () { return client_1.CoinbasePrimeClient; } });
var clientWithServices_1 = require("./clientWithServices");
Object.defineProperty(exports, "CoinbasePrimeClientWithServices", { enumerable: true, get: function () { return clientWithServices_1.CoinbasePrimeClientWithServices; } });
// Re-export core-ts objects for central control of service's usage of core-ts
var core_ts_1 = require("@coinbase-sample/core-ts");
Object.defineProperty(exports, "CoinbaseClient", { enumerable: true, get: function () { return core_ts_1.CoinbaseClient; } });
Object.defineProperty(exports, "Method", { enumerable: true, get: function () { return core_ts_1.Method; } });
Object.defineProperty(exports, "CoinbaseClientException", { enumerable: true, get: function () { return core_ts_1.CoinbaseClientException; } });
Object.defineProperty(exports, "CoinbaseError", { enumerable: true, get: function () { return core_ts_1.CoinbaseError; } });
