// @coinbase-sample/core-ts/dist/error/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinbaseClientException = exports.CoinbaseError = void 0;
exports.handleException = handleException;
class CoinbaseError extends Error {
    statusCode;
    response;
    constructor(message, statusCode, response) {
        super(message);
        this.name = 'CoinbaseError';
        this.statusCode = statusCode;
        this.response = response;
    }
}
exports.CoinbaseError = CoinbaseError;
class CoinbaseClientException extends Error {
    constructor(message) {
        super(message);
        this.name = 'CoinbaseClientException';
    }
}
exports.CoinbaseClientException = CoinbaseClientException;
function handleException(response, responseText, reason) {
    let message;
    if (response?.status && 400 <= response.status && response.status <= 599) {
        if (response.status == 403 &&
            responseText?.includes &&
            responseText.includes('"error_details":"Missing required scopes"')) {
            message = `${response.status} Coinbase Error: Missing Required Scopes. Please verify your API keys include the necessary permissions.`;
        }
        else if (response.status == 400) {
            message = `${response.status} Coinbase Invalid Request Error: ${reason} ${response?.data?.message}`;
        }
        else
            message = `${response.status} Coinbase Error: ${reason}`;
        throw new CoinbaseError(message, response.status, response);
    }
}
