// @coinbase-sample/prime-sdk-ts/dist/model/enumPrefixes.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * This file is auto-generated during the type generation process.
 * Do not edit manually - regenerate by running: npm run generate-types
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENUM_ANALYSIS = exports.GENERATED_ENUM_PREFIXES = void 0;
/**
 * Auto-generated enum prefix patterns extracted from OpenAPI specification
 * Total enums analyzed: 43
 * Prefix patterns discovered: 77
 */
exports.GENERATED_ENUM_PREFIXES = [
    'ACTION_',
    'ACTIVITY_CATEGORY_',
    'ACTIVITY_LEVEL_',
    'ACTIVITY_SECONDARY_',
    'ACTIVITY_STATUS_',
    'ACTIVITY_TYPE_',
    'ADDRESS_BOOK_',
    'ALLOCATION_STATUS_',
    'BALANCE_',
    'BENCHMARK_',
    'BILATERAL_',
    'BILLING_',
    'CHANGE_',
    'CLAIM_',
    'COINBASE_',
    'COMPLETE_',
    'CONCENTRATION_',
    'COUNTERPARTY_',
    'CRYPTO_',
    'DEPOSIT_',
    'DESTINATION_',
    'DESTINATION_PAYMENT_',
    'FCM_FUTURES_',
    'FCM_MARGIN_',
    'FCM_POSITION_',
    'FILL_OR_',
    'FULL_',
    'GOOD_UNTIL_',
    'HIERARCHY_TYPE_',
    'IMMEDIATE_OR_',
    'INTERNAL_',
    'INVOICE_STATE_',
    'INVOICE_TYPE_',
    'ITEM_',
    'ITEM_APPROVAL_',
    'LOAN_TYPE_',
    'MACRO_',
    'MARGIN_ADD_',
    'METHOD_',
    'MULTIPLE_',
    'NETWORK_FAMILY_',
    'NETWORK_TYPE_',
    'NO_SECONDARY_',
    'ONCHAIN_',
    'OTHER_',
    'OTHER_ACTIVITY_',
    'OTHER_TRANSACTION_',
    'PAYMENT_',
    'PORTFOLIO_',
    'POSITION_REFERENCE_',
    'PRIME_CUSTODY_',
    'PRODUCT_PERMISSION_',
    'PROXY_',
    'RATE_TYPE_',
    'REMOVE_AUTHORIZED_',
    'SHORT_',
    'SHORT_BIASED_',
    'SHORT_COLLATERAL_',
    'SINGLE_COIN_',
    'STAKE_ACCOUNT_',
    'STAKE_AUTHORIZE_',
    'STOP_',
    'SWEEP_',
    'TEAM_',
    'TOTAL_',
    'TRADE_',
    'TRADING_',
    'TRANSACTION_',
    'TRANSACTION_IMPORT_',
    'TRANSACTION_TYPE_',
    'UNIFIED_TOTAL_',
    'VAULT_',
    'VOTE_',
    'WALLET_TYPE_',
    'WALLET_VISIBILITY_',
    'WITHDRAWAL_',
    'WITHDRAW_',
];
/**
 * Detailed enum analysis for debugging and insights
 */
exports.ENUM_ANALYSIS = {
    totalEnums: 43,
    discoveredPrefixes: 77,
    prefixPatterns: exports.GENERATED_ENUM_PREFIXES,
    enumDetails: [
        {
            name: 'CoinbaseCustodyApiActivityType',
            valueCount: 29,
            file: 'coinbaseCustodyApiActivityType.ts',
        },
        {
            name: 'CoinbaseCustodyApiAddressBookType',
            valueCount: 3,
            file: 'coinbaseCustodyApiAddressBookType.ts',
        },
        {
            name: 'CoinbasePublicRestApiAction',
            valueCount: 5,
            file: 'coinbasePublicRestApiAction.ts',
        },
        {
            name: 'CoinbasePublicRestApiActivityCategory',
            valueCount: 7,
            file: 'coinbasePublicRestApiActivityCategory.ts',
        },
        {
            name: 'CoinbasePublicRestApiActivityLevel',
            valueCount: 3,
            file: 'coinbasePublicRestApiActivityLevel.ts',
        },
        {
            name: 'CoinbasePublicRestApiActivitySecondaryType',
            valueCount: 7,
            file: 'coinbasePublicRestApiActivitySecondaryType.ts',
        },
        {
            name: 'CoinbasePublicRestApiActivityStatus',
            valueCount: 7,
            file: 'coinbasePublicRestApiActivityStatus.ts',
        },
        {
            name: 'CoinbasePublicRestApiActivityType',
            valueCount: 42,
            file: 'coinbasePublicRestApiActivityType.ts',
        },
        {
            name: 'CoinbasePublicRestApiAllocationSizeType',
            valueCount: 3,
            file: 'coinbasePublicRestApiAllocationSizeType.ts',
        },
        {
            name: 'CoinbasePublicRestApiAllocationStatus',
            valueCount: 5,
            file: 'coinbasePublicRestApiAllocationStatus.ts',
        },
        {
            name: 'CoinbasePublicRestApiAssetChangeType',
            valueCount: 5,
            file: 'coinbasePublicRestApiAssetChangeType.ts',
        },
        {
            name: 'CoinbasePublicRestApiBenchmark',
            valueCount: 5,
            file: 'coinbasePublicRestApiBenchmark.ts',
        },
        {
            name: 'CoinbasePublicRestApiDestinationType',
            valueCount: 4,
            file: 'coinbasePublicRestApiDestinationType.ts',
        },
        {
            name: 'CoinbasePublicRestApiFcmFuturesSweepStatus',
            valueCount: 5,
            file: 'coinbasePublicRestApiFcmFuturesSweepStatus.ts',
        },
        {
            name: 'CoinbasePublicRestApiFcmMarginCallState',
            valueCount: 5,
            file: 'coinbasePublicRestApiFcmMarginCallState.ts',
        },
        {
            name: 'CoinbasePublicRestApiFcmMarginCallType',
            valueCount: 3,
            file: 'coinbasePublicRestApiFcmMarginCallType.ts',
        },
        {
            name: 'CoinbasePublicRestApiFcmPositionSide',
            valueCount: 3,
            file: 'coinbasePublicRestApiFcmPositionSide.ts',
        },
        {
            name: 'CoinbasePublicRestApiHierarchyType',
            valueCount: 3,
            file: 'coinbasePublicRestApiHierarchyType.ts',
        },
        {
            name: 'CoinbasePublicRestApiInvoiceState',
            valueCount: 5,
            file: 'coinbasePublicRestApiInvoiceState.ts',
        },
        {
            name: 'CoinbasePublicRestApiInvoiceType',
            valueCount: 6,
            file: 'coinbasePublicRestApiInvoiceType.ts',
        },
        {
            name: 'CoinbasePublicRestApiLoanType',
            valueCount: 6,
            file: 'coinbasePublicRestApiLoanType.ts',
        },
        {
            name: 'CoinbasePublicRestApiMarginAddOnType',
            valueCount: 5,
            file: 'coinbasePublicRestApiMarginAddOnType.ts',
        },
        {
            name: 'CoinbasePublicRestApiNetworkFamily',
            valueCount: 3,
            file: 'coinbasePublicRestApiNetworkFamily.ts',
        },
        {
            name: 'CoinbasePublicRestApiNetworkType',
            valueCount: 3,
            file: 'coinbasePublicRestApiNetworkType.ts',
        },
        {
            name: 'CoinbasePublicRestApiOrderSide',
            valueCount: 2,
            file: 'coinbasePublicRestApiOrderSide.ts',
        },
        {
            name: 'CoinbasePublicRestApiOrderStatus',
            valueCount: 6,
            file: 'coinbasePublicRestApiOrderStatus.ts',
        },
        {
            name: 'CoinbasePublicRestApiOrderType',
            valueCount: 7,
            file: 'coinbasePublicRestApiOrderType.ts',
        },
        {
            name: 'CoinbasePublicRestApiPaymentMethodType',
            valueCount: 3,
            file: 'coinbasePublicRestApiPaymentMethodType.ts',
        },
        {
            name: 'CoinbasePublicRestApiPortfolioBalanceType',
            valueCount: 5,
            file: 'coinbasePublicRestApiPortfolioBalanceType.ts',
        },
        {
            name: 'CoinbasePublicRestApiPositionReferenceType',
            valueCount: 3,
            file: 'coinbasePublicRestApiPositionReferenceType.ts',
        },
        {
            name: 'CoinbasePublicRestApiProductPermissions',
            valueCount: 3,
            file: 'coinbasePublicRestApiProductPermissions.ts',
        },
        {
            name: 'CoinbasePublicRestApiRateType',
            valueCount: 5,
            file: 'coinbasePublicRestApiRateType.ts',
        },
        {
            name: 'CoinbasePublicRestApiSigningStatus',
            valueCount: 2,
            file: 'coinbasePublicRestApiSigningStatus.ts',
        },
        {
            name: 'CoinbasePublicRestApiSortDirection',
            valueCount: 2,
            file: 'coinbasePublicRestApiSortDirection.ts',
        },
        {
            name: 'CoinbasePublicRestApiTimeInForceType',
            valueCount: 4,
            file: 'coinbasePublicRestApiTimeInForceType.ts',
        },
        {
            name: 'CoinbasePublicRestApiTransactionStatus',
            valueCount: 21,
            file: 'coinbasePublicRestApiTransactionStatus.ts',
        },
        {
            name: 'CoinbasePublicRestApiTransactionType',
            valueCount: 33,
            file: 'coinbasePublicRestApiTransactionType.ts',
        },
        {
            name: 'CoinbasePublicRestApiTransferLocationType',
            valueCount: 6,
            file: 'coinbasePublicRestApiTransferLocationType.ts',
        },
        {
            name: 'CoinbasePublicRestApiUserRole',
            valueCount: 9,
            file: 'coinbasePublicRestApiUserRole.ts',
        },
        {
            name: 'CoinbasePublicRestApiVisibilityStatus',
            valueCount: 3,
            file: 'coinbasePublicRestApiVisibilityStatus.ts',
        },
        {
            name: 'CoinbasePublicRestApiWalletDepositInstructionType',
            valueCount: 5,
            file: 'coinbasePublicRestApiWalletDepositInstructionType.ts',
        },
        {
            name: 'CoinbasePublicRestApiWalletType',
            valueCount: 5,
            file: 'coinbasePublicRestApiWalletType.ts',
        },
        {
            name: 'CoinbasePublicRestApiWalletVisibility',
            valueCount: 3,
            file: 'coinbasePublicRestApiWalletVisibility.ts',
        },
    ],
};
