// @coinbase-sample/core-ts/dist/client/index.js
"use strict";
/**
 * Copyright 2024-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinbaseClient = void 0;
const constants_1 = require("../constants");
const httpClient_1 = require("../http/httpClient");
class CoinbaseClient {
    apiBasePath;
    userAgent;
    httpClient;
    constructor(apiBasePath, credentials, userAgent, options) {
        this.apiBasePath = apiBasePath;
        if (typeof userAgent === 'string' && userAgent.length > 0)
            this.userAgent = userAgent;
        else
            this.userAgent = constants_1.USER_AGENT;
        this.httpClient = new httpClient_1.CoinbaseHttpClient(apiBasePath, this.userAgent, credentials, options);
    }
    request(options) {
        return this.httpClient.sendRequest(options);
    }
    addHeader(key, value) {
        this.httpClient.addHeader(key, value);
    }
    addTransformRequest(func) {
        this.httpClient.addTransformRequest(func);
    }
    addTransformResponse(func) {
        this.httpClient.addTransformResponse(func);
    }
    getDefaultPaginationLimit() {
        return this.httpClient.getDefaultPaginationLimit();
    }
    getMaxPages() {
        return this.httpClient.getMaxPages();
    }
    getMaxItems() {
        return this.httpClient.getMaxItems();
    }
}
exports.CoinbaseClient = CoinbaseClient;
