// @coinbase-sample/prime-sdk-ts/dist/shared/enumValidationCore.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamicEnumValidationError = void 0;
exports.isValueValidForEnum = isValueValidForEnum;
exports.detectEnumType = detectEnumType;
exports.analyzeField = analyzeField;
exports.validateEnumValue = validateEnumValue;
/**
 * Core validation logic for enum fields
 */
const enumHelpers_1 = require("./enumHelpers");
const enumPrefixes_1 = require("../model/enumPrefixes");
const enumRegistry_1 = require("./enumRegistry");
const fieldMapping_1 = require("./fieldMapping");
// Sort prefixes by length (longest first) to prevent collision issues
const ENUM_PREFIXES = [...enumPrefixes_1.GENERATED_ENUM_PREFIXES].sort((a, b) => b.length - a.length);
/**
 * Error thrown when a request contains invalid enum values
 */
class DynamicEnumValidationError extends Error {
    constructor(field, value, enumName, validValues, validKeys) {
        const validValuesStr = validValues.join(', ');
        const validKeysStr = validKeys
            ? ` Valid keys: ${validKeys.join(', ')}.`
            : '';
        super(`Invalid ${field} (${enumName}): "${value}". Valid values: ${validValuesStr}.${validKeysStr} ` +
            `Consider using enum helpers for automatic normalization.`);
        this.field = field;
        this.value = value;
        this.enumName = enumName;
        this.validValues = validValues;
        this.validKeys = validKeys;
        this.name = 'DynamicEnumValidationError';
    }
}
exports.DynamicEnumValidationError = DynamicEnumValidationError;
/**
 * Check if a value is valid for a specific enum (includes normalization)
 */
function isValueValidForEnum(value, enumObject) {
    if (!value || !enumObject)
        return false;
    // Direct value match
    if (Object.values(enumObject).includes(value))
        return true;
    // Direct key match
    if (Object.keys(enumObject).includes(value))
        return true;
    // Normalized value matches
    const normalizedValue = (0, enumHelpers_1.normalizeEnumValue)(value);
    if (Object.values(enumObject).includes(normalizedValue))
        return true;
    // Check with prefix normalization
    for (const prefix of ENUM_PREFIXES) {
        const withPrefix = `${prefix}${normalizedValue}`;
        if (Object.values(enumObject).includes(withPrefix))
            return true;
    }
    return false;
}
/**
 * Detects the enum type for a field based on field name patterns and value analysis
 */
function detectEnumType(fieldName, value, serviceName) {
    // Strategy 1: Direct field name to enum name mapping (with service context)
    const directMapping = (0, fieldMapping_1.getDirectEnumMapping)(fieldName, serviceName);
    if (directMapping) {
        const enumObject = enumRegistry_1.enumRegistry.getEnum(directMapping);
        if (enumObject) {
            // If we have a confident field mapping, use it even if value doesn't match
            // This allows validation to catch invalid values
            return { enum: enumObject, name: directMapping };
        }
    }
    // Strategy 2: Find enum by value (exact match)
    const enumByValue = enumRegistry_1.enumRegistry.findEnumByValue(value);
    if (enumByValue) {
        return enumByValue;
    }
    // Strategy 3: Find enum by value with prefix normalization
    const enumByValueWithPrefix = findEnumWithPrefixHandling(value);
    if (enumByValueWithPrefix) {
        return enumByValueWithPrefix;
    }
    // Strategy 4: Find enum by key (treating value as a potential key)
    const enumByKey = enumRegistry_1.enumRegistry.findEnumByKey(value);
    if (enumByKey) {
        return enumByKey;
    }
    return null;
}
/**
 * Find enum by value with automatic prefix handling
 */
function findEnumWithPrefixHandling(value) {
    const normalizedValue = (0, enumHelpers_1.normalizeEnumValue)(value);
    // Try each prefix
    for (const prefix of ENUM_PREFIXES) {
        const prefixedValue = `${prefix}${normalizedValue}`;
        const enumByPrefixedValue = enumRegistry_1.enumRegistry.findEnumByValue(prefixedValue);
        if (enumByPrefixedValue) {
            return enumByPrefixedValue;
        }
    }
    return null;
}
/**
 * Analyzes a single field to determine if it's an enum
 */
function analyzeField(fieldName, value, serviceName) {
    if (value === undefined || value === null) {
        return null;
    }
    // Handle array values
    if (Array.isArray(value)) {
        const firstNonNullValue = value.find((v) => v !== null && v !== undefined);
        if (firstNonNullValue && typeof firstNonNullValue === 'string') {
            const enumInfo = detectEnumType(fieldName, firstNonNullValue, serviceName);
            if (enumInfo) {
                return {
                    fieldName,
                    enumName: enumInfo.name,
                    isOptional: false,
                    isArray: true,
                };
            }
        }
        return null;
    }
    // Handle string values (potential enums)
    if (typeof value === 'string') {
        const enumInfo = detectEnumType(fieldName, value, serviceName);
        if (enumInfo) {
            return {
                fieldName,
                enumName: enumInfo.name,
                isOptional: false,
                isArray: false,
            };
        }
    }
    return null;
}
/**
 * Validate and normalize a single enum value
 */
function validateEnumValue(value, enumObject, enumName, fieldName, config) {
    // Direct value match
    if (Object.values(enumObject).includes(value)) {
        return { isValid: true, normalizedValue: value };
    }
    // Direct key match (convert key to value)
    const keyEntry = Object.entries(enumObject).find(([key]) => key === value);
    if (keyEntry) {
        return { isValid: true, normalizedValue: keyEntry[1] };
    }
    if (config.autoNormalize) {
        // Try normalized value
        const normalizedValue = (0, enumHelpers_1.normalizeEnumValue)(value);
        if (Object.values(enumObject).includes(normalizedValue)) {
            return { isValid: true, normalizedValue };
        }
        // Try with prefix normalization
        for (const prefix of ENUM_PREFIXES) {
            const withPrefix = `${prefix}${normalizedValue}`;
            if (Object.values(enumObject).includes(withPrefix)) {
                return { isValid: true, normalizedValue: withPrefix };
            }
        }
        // Try case-insensitive key match
        const normalizedKey = Object.keys(enumObject).find((key) => key.toLowerCase() === value.toLowerCase());
        if (normalizedKey) {
            return { isValid: true, normalizedValue: enumObject[normalizedKey] };
        }
    }
    // Value is invalid
    const error = `Invalid ${fieldName} (${enumName}): "${value}". Valid values: ${Object.values(enumObject).join(', ')}. Valid keys: ${Object.keys(enumObject).join(', ')}.`;
    return { isValid: false, error };
}
