// @coinbase-sample/prime-sdk-ts/dist/shared/enumValidators.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.toEnumFormat = exports.TransactionEnums = exports.OrderEnums = void 0;
exports.createEnumValidators = createEnumValidators;
/**
 * Convenient enum validators for common Prime API enums.
 * These provide pre-configured validation for frequently used enum types.
 *
 * Note: Most users should use the dynamic validation system instead, which
 * automatically handles all enums. These validators are useful for explicit
 * control or when you need to validate individual enum values.
 */
const enumHelpers_1 = require("./enumHelpers");
Object.defineProperty(exports, "toEnumFormat", { enumerable: true, get: function () { return enumHelpers_1.normalizeEnumValue; } });
// Import specific enums for validation
const OrderType_1 = require("../model/enums/OrderType");
const OrderSide_1 = require("../model/enums/OrderSide");
const OrderStatus_1 = require("../model/enums/OrderStatus");
const TransactionType_1 = require("../model/enums/TransactionType");
const TimeInForceType_1 = require("../model/enums/TimeInForceType");
/**
 * Order-related enum helpers
 */
exports.OrderEnums = {
    /**
     * Validates and normalizes order type values
     * Accepts: 'market', 'limit', 'twap', 'block', 'vwap', 'stopLimit', 'rfq' (case-insensitive)
     */
    getOrderType: (input) => (0, enumHelpers_1.getEnumValue)(OrderType_1.OrderType, input, 'OrderType'),
    requireOrderType: (input) => (0, enumHelpers_1.requireEnumValue)(OrderType_1.OrderType, input, 'OrderType'),
    isValidOrderType: (input) => (0, enumHelpers_1.isValidEnumValue)(OrderType_1.OrderType, input),
    /**
     * Validates and normalizes order side values
     * Accepts: 'buy', 'sell' (case-insensitive)
     */
    getOrderSide: (input) => (0, enumHelpers_1.getEnumValue)(OrderSide_1.OrderSide, input, 'OrderSide'),
    requireOrderSide: (input) => (0, enumHelpers_1.requireEnumValue)(OrderSide_1.OrderSide, input, 'OrderSide'),
    isValidOrderSide: (input) => (0, enumHelpers_1.isValidEnumValue)(OrderSide_1.OrderSide, input),
    /**
     * Validates and normalizes order status values
     * Accepts: 'open', 'filled', 'cancelled', 'expired', 'failed', 'pending' (case-insensitive)
     */
    getOrderStatus: (input) => (0, enumHelpers_1.getEnumValue)(OrderStatus_1.OrderStatus, input, 'OrderStatus'),
    requireOrderStatus: (input) => (0, enumHelpers_1.requireEnumValue)(OrderStatus_1.OrderStatus, input, 'OrderStatus'),
    isValidOrderStatus: (input) => (0, enumHelpers_1.isValidEnumValue)(OrderStatus_1.OrderStatus, input),
    /**
     * Validates and normalizes time in force type values
     * Accepts: 'good_until_cancelled', 'good_until_time', 'immediate_or_cancel', 'fill_or_kill' (case-insensitive)
     */
    getTimeInForceType: (input) => (0, enumHelpers_1.getEnumValue)(TimeInForceType_1.TimeInForceType, input, 'TimeInForceType'),
    requireTimeInForceType: (input) => (0, enumHelpers_1.requireEnumValue)(TimeInForceType_1.TimeInForceType, input, 'TimeInForceType'),
    isValidTimeInForceType: (input) => (0, enumHelpers_1.isValidEnumValue)(TimeInForceType_1.TimeInForceType, input),
};
/**
 * Transaction-related enum helpers
 */
exports.TransactionEnums = {
    /**
     * Validates and normalizes transaction type values
     * Accepts various transaction types like 'deposit', 'withdrawal', 'internalDeposit', etc. (case-insensitive)
     */
    getTransactionType: (input) => (0, enumHelpers_1.getEnumValue)(TransactionType_1.TransactionType, input, 'TransactionType'),
    requireTransactionType: (input) => (0, enumHelpers_1.requireEnumValue)(TransactionType_1.TransactionType, input, 'TransactionType'),
    isValidTransactionType: (input) => (0, enumHelpers_1.isValidEnumValue)(TransactionType_1.TransactionType, input),
};
/**
 * Utility for creating custom enum validators for any enum type
 *
 * @param enumObject - The enum object to create validators for
 * @param enumName - Name of the enum for error messages
 * @returns Object with validation methods
 *
 * @example
 * import { WalletType } from '../model/enums/WalletType';
 *
 * const WalletValidators = createEnumValidators(WalletType, 'WalletType');
 * WalletValidators.get('trading'); // Gets WalletType value
 * WalletValidators.require('custody'); // Gets WalletType value or throws
 * WalletValidators.isValid('invalid'); // false
 */
function createEnumValidators(enumObject, enumName) {
    return {
        get: (input) => (0, enumHelpers_1.getEnumValue)(enumObject, input, enumName),
        require: (input) => (0, enumHelpers_1.requireEnumValue)(enumObject, input, enumName),
        isValid: (input) => (0, enumHelpers_1.isValidEnumValue)(enumObject, input),
        values: Object.values(enumObject),
        keys: Object.keys(enumObject),
    };
}
/**
 * @deprecated Use the dynamic validation system instead.
 *
 * Most users should prefer:
 * ```javascript
 * import { dynamicValidateRequest } from '@coinbase/prime-sdk';
 * const validated = dynamicValidateRequest(request);
 * ```
 *
 * This provides automatic validation for all enum fields without manual configuration.
 */
