// @coinbase-sample/prime-sdk-ts/dist/shared/enumRegistry.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.enumRegistry = exports.EnumRegistry = void 0;
/**
 * Central registry for managing and discovering TypeScript enums
 */
// Import all enums dynamically
const AllEnums = __importStar(require("../model/enums/"));
/**
 * Registry for managing discovered enums and providing lookup capabilities
 */
class EnumRegistry {
    constructor() {
        this.enums = new Map();
        this.enumsByValue = new Map();
        this.enumsByKey = new Map();
        this.populateEnums();
    }
    /**
     * Automatically discover and register all enums from the model/enums directory
     */
    populateEnums() {
        Object.entries(AllEnums).forEach(([enumName, enumObject]) => {
            if (this.isValidEnum(enumObject)) {
                this.enums.set(enumName, enumObject);
                this.indexEnumValues(enumName, enumObject);
            }
        });
    }
    /**
     * Check if an object is a valid TypeScript enum
     */
    isValidEnum(obj) {
        return (obj &&
            typeof obj === 'object' &&
            Object.values(obj).every((value) => typeof value === 'string'));
    }
    /**
     * Index enum values and keys for fast lookup
     */
    indexEnumValues(enumName, enumObject) {
        Object.entries(enumObject).forEach(([key, value]) => {
            // Index by value (e.g., "BUY" -> OrderSide)
            this.enumsByValue.set(value.toLowerCase(), {
                enum: enumObject,
                name: enumName,
            });
            // Index by key (e.g., "Buy" -> OrderSide)
            this.enumsByKey.set(key.toLowerCase(), {
                enum: enumObject,
                name: enumName,
            });
        });
    }
    /**
     * Get a specific enum by name
     */
    getEnum(enumName) {
        return this.enums.get(enumName) || null;
    }
    /**
     * Get all registered enum names
     */
    getAllEnumNames() {
        return Array.from(this.enums.keys());
    }
    /**
     * Find enum by value (exact match)
     */
    findEnumByValue(value) {
        return this.enumsByValue.get(value.toLowerCase()) || null;
    }
    /**
     * Find enum by key (exact match)
     */
    findEnumByKey(key) {
        return this.enumsByKey.get(key.toLowerCase()) || null;
    }
    /**
     * Get statistics about discovered enums
     */
    getStats() {
        return {
            totalEnums: this.enums.size,
            totalValues: this.enumsByValue.size,
            totalKeys: this.enumsByKey.size,
        };
    }
    /**
     * Get detailed information about all enums
     */
    getEnumDetails() {
        const details = {};
        this.enums.forEach((enumObject, enumName) => {
            details[enumName] = {
                values: Object.values(enumObject),
                keys: Object.keys(enumObject),
            };
        });
        return details;
    }
}
exports.EnumRegistry = EnumRegistry;
// Global singleton instance
exports.enumRegistry = new EnumRegistry();
