// @coinbase-sample/prime-sdk-ts/dist/shared/enumHelpers.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeEnumValue = normalizeEnumValue;
exports.getEnumValue = getEnumValue;
exports.requireEnumValue = requireEnumValue;
exports.isValidEnumValue = isValidEnumValue;
exports.getEnumValues = getEnumValues;
exports.getEnumKeys = getEnumKeys;
exports.validateEnumPrefix = validateEnumPrefix;
exports.transformEnumValue = transformEnumValue;
/**
 * Enum transformation utilities for JavaScript developers.
 * These helpers make it easier to work with enums by providing:
 * - Case-insensitive enum lookups
 * - CamelCase to SNAKE_CASE conversion
 * - Enum validation with helpful error messages
 * - Support for prefix-aware transformations
 */
/**
 * Converts a string from various formats to the expected enum value format (SCREAMING_SNAKE_CASE)
 *
 * @param input - The input string in any case format
 * @returns The normalized SCREAMING_SNAKE_CASE string
 *
 * @example
 * normalizeEnumValue('buy') // 'BUY'
 * normalizeEnumValue('marketOrder') // 'MARKET_ORDER'
 * normalizeEnumValue('STOP_LIMIT') // 'STOP_LIMIT'
 * normalizeEnumValue('Order_Type_Market') // 'ORDER_TYPE_MARKET'
 */
function normalizeEnumValue(input) {
    if (!input || typeof input !== 'string') {
        throw new Error('Input must be a non-empty string');
    }
    return (input
        // Convert camelCase to snake_case
        .replace(/([a-z])([A-Z])/g, '$1_$2')
        // Handle existing underscores and hyphens
        .replace(/[-\s]+/g, '_')
        // Convert to uppercase
        .toUpperCase()
        // Clean up multiple underscores
        .replace(/_+/g, '_')
        // Remove leading/trailing underscores
        .replace(/^_+|_+$/g, ''));
}
/**
 * Safely gets an enum value by normalizing the input and checking if it exists
 *
 * @param enumObject - The enum object to search in
 * @param input - The input value to normalize and find
 * @param enumName - Optional enum name for better error messages
 * @returns The matching enum value or undefined if not found
 *
 * @example
 * import { OrderSide } from '../model/enums/OrderSide';
 *
 * getEnumValue(OrderSide, 'buy') // OrderSide.Buy
 * getEnumValue(OrderSide, 'BUY') // OrderSide.Buy
 * getEnumValue(OrderSide, 'invalid') // undefined
 */
function getEnumValue(enumObject, input, enumName) {
    if (!input || typeof input !== 'string') {
        return undefined;
    }
    const normalizedInput = normalizeEnumValue(input);
    // Direct value match (input is already the enum value)
    if (Object.values(enumObject).includes(normalizedInput)) {
        return normalizedInput;
    }
    // Key-based lookup (find enum key that matches)
    for (const [key, value] of Object.entries(enumObject)) {
        if (normalizeEnumValue(key) === normalizedInput) {
            return value;
        }
    }
    return undefined;
}
/**
 * Gets an enum value with strict validation, throwing an error if not found
 *
 * @param enumObject - The enum object to search in
 * @param input - The input value to normalize and find
 * @param enumName - Optional enum name for better error messages
 * @returns The matching enum value
 * @throws Error if the enum value is not found
 *
 * @example
 * import { OrderSide } from '../model/enums/OrderSide';
 *
 * requireEnumValue(OrderSide, 'buy') // OrderSide.Buy
 * requireEnumValue(OrderSide, 'invalid') // throws Error
 */
function requireEnumValue(enumObject, input, enumName) {
    const result = getEnumValue(enumObject, input, enumName);
    if (result === undefined) {
        const enumNameStr = enumName || 'Enum';
        const validValues = Object.values(enumObject).join(', ');
        const validKeys = Object.keys(enumObject).join(', ');
        throw new Error(`Invalid ${enumNameStr} value: "${input}". ` +
            `Valid values: ${validValues}. ` +
            `Valid keys: ${validKeys}.`);
    }
    return result;
}
/**
 * Checks if a string is a valid enum value (case-insensitive)
 *
 * @param enumObject - The enum object to check against
 * @param input - The input value to validate
 * @returns True if the input is a valid enum value
 *
 * @example
 * import { OrderSide } from '../model/enums/OrderSide';
 *
 * isValidEnumValue(OrderSide, 'buy') // true
 * isValidEnumValue(OrderSide, 'BUY') // true
 * isValidEnumValue(OrderSide, 'invalid') // false
 */
function isValidEnumValue(enumObject, input) {
    return getEnumValue(enumObject, input) !== undefined;
}
/**
 * Gets all possible enum values as an array
 *
 * @param enumObject - The enum object
 * @returns Array of all enum values
 *
 * @example
 * import { OrderSide } from '../model/enums/OrderSide';
 *
 * getEnumValues(OrderSide) // ['BUY', 'SELL']
 */
function getEnumValues(enumObject) {
    return Object.values(enumObject);
}
/**
 * Gets all possible enum keys as an array
 *
 * @param enumObject - The enum object
 * @returns Array of all enum keys
 *
 * @example
 * import { OrderSide } from '../model/enums/OrderSide';
 *
 * getEnumKeys(OrderSide) // ['Buy', 'Sell']
 */
function getEnumKeys(enumObject) {
    return Object.keys(enumObject);
}
/**
 * Validates that a string has the correct prefix for enum values
 * Useful for ensuring enum values follow expected patterns
 *
 * @param input - The enum value to check
 * @param expectedPrefix - The expected prefix (e.g., 'ORDER_TYPE_', 'ACTIVITY_TYPE_')
 * @param strict - If true, requires exact prefix match. If false, allows values without prefix
 * @returns True if the prefix is correct
 *
 * @example
 * validateEnumPrefix('ORDER_TYPE_MARKET', 'ORDER_TYPE_') // true
 * validateEnumPrefix('MARKET', 'ORDER_TYPE_', false) // true (non-strict)
 * validateEnumPrefix('MARKET', 'ORDER_TYPE_', true) // false (strict)
 */
function validateEnumPrefix(input, expectedPrefix, strict = false) {
    if (!input || !expectedPrefix) {
        return false;
    }
    const normalizedInput = normalizeEnumValue(input);
    const normalizedPrefix = normalizeEnumValue(expectedPrefix);
    if (strict) {
        return normalizedInput.startsWith(normalizedPrefix);
    }
    // Non-strict: allow values with or without prefix
    return (normalizedInput.startsWith(normalizedPrefix) ||
        !normalizedInput.includes('_') ||
        normalizedInput ===
            normalizeEnumValue(input.replace(new RegExp(`^${expectedPrefix}`, 'i'), '')));
}
/**
 * Transforms enum input to the correct format, optionally adding a prefix
 *
 * @param input - The input string
 * @param prefix - Optional prefix to add if not present
 * @returns The transformed enum value
 *
 * @example
 * transformEnumValue('market', 'ORDER_TYPE_') // 'ORDER_TYPE_MARKET'
 * transformEnumValue('ORDER_TYPE_MARKET', 'ORDER_TYPE_') // 'ORDER_TYPE_MARKET'
 * transformEnumValue('marketOrder') // 'MARKET_ORDER'
 */
function transformEnumValue(input, prefix) {
    const normalized = normalizeEnumValue(input);
    if (!prefix) {
        return normalized;
    }
    const normalizedPrefix = normalizeEnumValue(prefix);
    // If already has the prefix, return as-is
    if (normalized.startsWith(normalizedPrefix)) {
        return normalized;
    }
    // Add prefix
    return `${normalizedPrefix}${normalized}`;
}
