// @coinbase-sample/prime-sdk-ts/dist/shared/fieldMapping.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFieldNameVariations = getFieldNameVariations;
exports.extractContextClues = extractContextClues;
exports.findEnumWithContext = findEnumWithContext;
exports.findEnumByFieldName = findEnumByFieldName;
exports.getDirectEnumMapping = getDirectEnumMapping;
/**
 * Field-to-enum mapping logic for dynamic enum detection
 */
const enumRegistry_1 = require("./enumRegistry");
const serviceContext_1 = require("./serviceContext");
/**
 * Generate field name variations for plural/singular matching
 */
function getFieldNameVariations(fieldName) {
    const variations = [fieldName];
    // Handle irregular plurals and common patterns
    const irregularPlurals = {
        categories: 'category',
        statuses: 'status',
        activities: 'activity',
        entities: 'entity',
        currencies: 'currency',
        portfolios: 'portfolio',
    };
    const irregularSingulars = {
        category: 'categories',
        status: 'statuses',
        activity: 'activities',
        entity: 'entities',
        currency: 'currencies',
        portfolio: 'portfolios',
    };
    // Check irregular mappings first
    if (irregularPlurals[fieldName]) {
        variations.push(irregularPlurals[fieldName]);
    }
    if (irregularSingulars[fieldName]) {
        variations.push(irregularSingulars[fieldName]);
    }
    // Handle regular -s plurals only if not already handled by irregular rules
    if (!irregularPlurals[fieldName] && !irregularSingulars[fieldName]) {
        if (fieldName.endsWith('s') && fieldName.length > 3) {
            // Remove 's' for potential singular
            const singular = fieldName.slice(0, -1);
            // Avoid cases like "address" -> "addres"
            if (!singular.endsWith('s')) {
                variations.push(singular);
            }
        }
        else {
            // Add regular plural
            variations.push(fieldName + 's');
        }
    }
    return variations;
}
/**
 * Extract context clues from field names to resolve enum ambiguity
 */
function extractContextClues(fieldName) {
    const clues = [];
    const fieldLower = fieldName.toLowerCase();
    // Extract prefix context (e.g., "activitystatus" → ["activity"])
    const prefixMatch = fieldLower.match(/^([a-z]+?)(?:status|type|category)$/);
    if (prefixMatch) {
        clues.push(prefixMatch[1]);
    }
    // Extract camelCase components (e.g., "activityStatus" → ["activity"])
    const camelComponents = fieldName
        .replace(/([A-Z])/g, '_$1')
        .toLowerCase()
        .split('_')
        .filter(Boolean);
    clues.push(...camelComponents.slice(0, -1)); // Exclude the last component (status/type/category)
    // Extract underscore/dash separated components
    const separatedComponents = fieldName.toLowerCase().split(/[_-]+/);
    clues.push(...separatedComponents.slice(0, -1));
    return [...new Set(clues)]; // Remove duplicates
}
/**
 * Find enum with context awareness to resolve collisions
 */
function findEnumWithContext(fieldName, allEnumNames) {
    // Handle ambiguous suffixes that appear in multiple enums
    const ambiguousSuffixes = ['status', 'type', 'category'];
    const fieldLower = fieldName.toLowerCase();
    // Check if this is an ambiguous field name
    const isAmbiguous = ambiguousSuffixes.some((suffix) => fieldLower.endsWith(suffix) || fieldLower === suffix);
    if (isAmbiguous) {
        // For ambiguous cases, look for context clues in the field name
        const contextClues = extractContextClues(fieldName);
        // Find enum that best matches the context
        const matches = allEnumNames.filter((enumName) => {
            const enumLower = enumName.toLowerCase();
            // Exact compound match (e.g., "activitystatus" → "ActivityStatus")
            if (enumLower === fieldLower)
                return true;
            // Context-based matching
            for (const clue of contextClues) {
                if (enumLower.includes(clue) &&
                    enumLower.endsWith(fieldLower.split(/(?=[a-z])/).pop() || '')) {
                    return true;
                }
            }
            return false;
        });
        // Return the most specific match
        if (matches.length === 1) {
            return matches[0];
        }
        // If multiple matches, prefer the longest/most specific one
        if (matches.length > 1) {
            return matches.reduce((longest, current) => current.length > longest.length ? current : longest);
        }
    }
    // Non-ambiguous cases: use careful substring matching
    // Sort enums by name length (shortest first) to prefer simpler matches
    const sortedEnumNames = allEnumNames
        .slice()
        .sort((a, b) => a.length - b.length);
    for (const enumName of sortedEnumNames) {
        const enumNameLower = enumName.toLowerCase();
        // Only match if field name is a meaningful part of the enum name
        // Avoid greedy matches like "side" matching "FcmPositionSide"
        if (enumNameLower === fieldLower) {
            return enumName; // Exact match
        }
        // Check if field name appears as a word boundary in enum name
        // e.g., "side" should match "OrderSide" but not "FcmPositionSide"
        const wordBoundaryRegex = new RegExp(`\\b${fieldLower}\\b`, 'i');
        if (wordBoundaryRegex.test(enumName)) {
            return enumName;
        }
        // Check for field name at the end of enum name (most common pattern)
        if (enumNameLower.endsWith(fieldLower) && enumNameLower !== fieldLower) {
            // Ensure it's not a substring within a word
            const beforeField = enumNameLower.slice(0, -fieldLower.length);
            if (beforeField.length > 0 && /[a-z]$/.test(beforeField)) {
                continue; // Skip if it's part of a word
            }
            return enumName;
        }
    }
    return null;
}
/**
 * Dynamically find enum name based on field name patterns
 */
function findEnumByFieldName(fieldName) {
    const allEnumNames = enumRegistry_1.enumRegistry.getAllEnumNames();
    // Strategy 1: Direct name match (e.g., "ActivityCategory" field → "ActivityCategory" enum)
    const exactMatch = allEnumNames.find((enumName) => enumName.toLowerCase() === fieldName.toLowerCase());
    if (exactMatch)
        return exactMatch;
    // Strategy 2: Smart plural/singular variations
    const variations = getFieldNameVariations(fieldName);
    for (const enumName of allEnumNames) {
        const enumNameLower = enumName.toLowerCase();
        if (variations.includes(enumNameLower)) {
            return enumName;
        }
    }
    // Strategy 3: Context-aware enum matching (handles collisions)
    const contextMatch = findEnumWithContext(fieldName, allEnumNames);
    if (contextMatch) {
        return contextMatch;
    }
    // Strategy 4: Smart pattern matching for compound field names
    // Handle cases like "activitycategory" → "ActivityCategory"
    for (const enumName of allEnumNames) {
        const enumWords = enumName
            .toLowerCase()
            .replace(/([A-Z])/g, ' $1')
            .trim()
            .split(' ');
        const fieldWords = fieldName
            .replace(/([A-Z])/g, ' $1')
            .trim()
            .split(/[\s_-]+/);
        // Check if all field words are contained in enum words
        const allWordsMatch = fieldWords.every((word) => enumWords.some((enumWord) => enumWord.includes(word) || word.includes(enumWord)));
        if (allWordsMatch && fieldWords.length > 1) {
            return enumName;
        }
    }
    return null;
}
/**
 * Get direct field-to-enum mapping with service context and fallback to essential mappings
 */
function getDirectEnumMapping(fieldName, serviceName) {
    const fieldNameLower = fieldName.toLowerCase();
    // First priority: Service context-specific mapping
    if (serviceName) {
        const serviceMapping = (0, serviceContext_1.resolveFieldWithServiceContext)(fieldName, serviceName);
        if (serviceMapping) {
            return serviceMapping;
        }
    }
    // Second priority: Dynamic enum name matching
    const dynamicMatch = findEnumByFieldName(fieldNameLower);
    if (dynamicMatch) {
        return dynamicMatch;
    }
    // Fall back to essential manual mappings for common patterns that don't follow naming conventions
    // These should be conservative and only include non-ambiguous cases
    const essentialMappings = {
        // Common trading/order fields (most frequent use case)
        side: 'OrderSide',
        status: 'OrderStatus', // Default to OrderStatus, validation will catch mismatches
        role: 'UserRole',
        action: 'Action',
        direction: 'SortDirection',
        // Specific compound fields to prevent wrong mappings
        orderside: 'OrderSide',
        ordertype: 'OrderType',
        orderstatus: 'OrderStatus',
        wallettype: 'WalletType',
        transactiontype: 'TransactionType',
        activitytype: 'CustodyActivityType',
        activitystatus: 'ActivityStatus',
        // Note: 'type' still removed (too ambiguous - let value-based detection handle it)
    };
    return essentialMappings[fieldNameLower] || null;
}
