// @coinbase-sample/prime-sdk-ts/dist/shared/serviceContext.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SERVICE_CONTEXTS = void 0;
exports.getServiceContext = getServiceContext;
exports.resolveFieldWithServiceContext = resolveFieldWithServiceContext;
exports.getAllServiceContexts = getAllServiceContexts;
exports.hasServiceContext = hasServiceContext;
/**
 * Registry of service contexts for accurate enum detection
 */
exports.SERVICE_CONTEXTS = {
    // Activities Service
    activities: {
        serviceName: 'ActivitiesService',
        fieldMappings: {
            status: 'ActivityStatus',
            type: 'CustodyActivityType',
            category: 'ActivityCategory',
            categories: 'ActivityCategory',
            secondaryType: 'ActivitySecondaryType',
        },
        description: 'Portfolio and entity activity operations',
    },
    // Orders Service
    orders: {
        serviceName: 'OrdersService',
        fieldMappings: {
            status: 'OrderStatus',
            type: 'OrderType',
            side: 'OrderSide',
            timeInForce: 'TimeInForceType',
        },
        description: 'Trading order operations',
    },
    // Transactions Service
    transactions: {
        serviceName: 'TransactionsService',
        fieldMappings: {
            status: 'TransactionStatus',
            type: 'TransactionType',
            destinationType: 'DestinationType',
        },
        description: 'Transaction and transfer operations',
    },
    // Wallets Service
    wallets: {
        serviceName: 'WalletsService',
        fieldMappings: {
            type: 'WalletType',
            status: 'WalletStatus',
        },
        description: 'Wallet management operations',
    },
    // Allocations Service
    allocations: {
        serviceName: 'AllocationsService',
        fieldMappings: {
            status: 'AllocationStatus',
            type: 'AllocationType',
        },
        description: 'Portfolio allocation operations',
    },
    // Users Service
    users: {
        serviceName: 'UsersService',
        fieldMappings: {
            role: 'UserRole',
            status: 'UserStatus',
        },
        description: 'User management operations',
    },
    // Balances Service
    balances: {
        serviceName: 'BalancesService',
        fieldMappings: {
            type: 'PortfolioBalanceType',
        },
        description: 'Balance and portfolio operations',
    },
    // Portfolios Service
    portfolios: {
        serviceName: 'PortfoliosService',
        fieldMappings: {
            type: 'PortfolioType',
            status: 'PortfolioStatus',
        },
        description: 'Portfolio management operations',
    },
    // Products Service
    products: {
        serviceName: 'ProductsService',
        fieldMappings: {
            type: 'ProductType',
            permissions: 'ProductPermissions',
            category: 'ProductCategory',
        },
        description: 'Product and market data operations',
    },
    // Assets Service
    assets: {
        serviceName: 'AssetsService',
        fieldMappings: {
            type: 'AssetType',
            status: 'AssetStatus',
        },
        description: 'Asset management operations',
    },
};
/**
 * Get service context by service name
 */
function getServiceContext(serviceName) {
    // Normalize service name (remove "Service" suffix if present)
    const normalizedName = serviceName.toLowerCase().replace(/service$/, '');
    return exports.SERVICE_CONTEXTS[normalizedName] || null;
}
/**
 * Resolve field to enum using service context
 */
function resolveFieldWithServiceContext(fieldName, serviceName) {
    if (!serviceName) {
        return null;
    }
    const context = getServiceContext(serviceName);
    if (!context) {
        return null;
    }
    const fieldLower = fieldName.toLowerCase();
    return context.fieldMappings[fieldLower] || null;
}
/**
 * Get all available service contexts
 */
function getAllServiceContexts() {
    return Object.assign({}, exports.SERVICE_CONTEXTS);
}
/**
 * Check if a service has context defined
 */
function hasServiceContext(serviceName) {
    return getServiceContext(serviceName) !== null;
}
