// @coinbase-sample/prime-sdk-ts/dist/model/enums/AssetChangeType.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * This file is generated by Openapi Generator https://github.com/openapitools/openapi-generator
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *  Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetChangeType = void 0;
var AssetChangeType;
(function (AssetChangeType) {
    AssetChangeType["BalanceTransfer"] = "BALANCE_TRANSFER";
    AssetChangeType["BalanceApproval"] = "BALANCE_APPROVAL";
    AssetChangeType["ItemTransfer"] = "ITEM_TRANSFER";
    AssetChangeType["ItemApproval"] = "ITEM_APPROVAL";
    AssetChangeType["ItemApprovalAll"] = "ITEM_APPROVAL_ALL";
})(AssetChangeType || (exports.AssetChangeType = AssetChangeType = {}));
