// @coinbase-sample/prime-sdk-ts/dist/model/enums/InvoiceType.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * This file is generated by Openapi Generator https://github.com/openapitools/openapi-generator
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *  Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvoiceType = void 0;
var InvoiceType;
(function (InvoiceType) {
    InvoiceType["InvoiceTypeUnspecified"] = "INVOICE_TYPE_UNSPECIFIED";
    InvoiceType["InvoiceTypeAucFee"] = "INVOICE_TYPE_AUC_FEE";
    InvoiceType["InvoiceTypeMinimumFee"] = "INVOICE_TYPE_MINIMUM_FEE";
    InvoiceType["InvoiceTypeWithdrawalFee"] = "INVOICE_TYPE_WITHDRAWAL_FEE";
    InvoiceType["InvoiceTypeNewWalletFee"] = "INVOICE_TYPE_NEW_WALLET_FEE";
    InvoiceType["InvoiceTypeStakingFee"] = "INVOICE_TYPE_STAKING_FEE";
})(InvoiceType || (exports.InvoiceType = InvoiceType = {}));
