// @coinbase-sample/prime-sdk-ts/dist/model/enums/PrimeActivityType.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * This file is generated by Openapi Generator https://github.com/openapitools/openapi-generator
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *  Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityType = void 0;
var ActivityType;
(function (ActivityType) {
    ActivityType["OtherActivityType"] = "OTHER_ACTIVITY_TYPE";
    ActivityType["ActivityTypeLimitOrder"] = "ACTIVITY_TYPE_LIMIT_ORDER";
    ActivityType["ActivityTypeMarketOrder"] = "ACTIVITY_TYPE_MARKET_ORDER";
    ActivityType["ActivityTypeTwapOrder"] = "ACTIVITY_TYPE_TWAP_ORDER";
    ActivityType["ActivityTypeBlockTrade"] = "ACTIVITY_TYPE_BLOCK_TRADE";
    ActivityType["ActivityTypeVwapOrder"] = "ACTIVITY_TYPE_VWAP_ORDER";
    ActivityType["ActivityTypeStopLimitOrder"] = "ACTIVITY_TYPE_STOP_LIMIT_ORDER";
    ActivityType["ActivityTypeDeposit"] = "ACTIVITY_TYPE_DEPOSIT";
    ActivityType["ActivityTypeWithdrawal"] = "ACTIVITY_TYPE_WITHDRAWAL";
    ActivityType["ActivityTypeInternalTransfer"] = "ACTIVITY_TYPE_INTERNAL_TRANSFER";
    ActivityType["ActivityTypeCreateWallet"] = "ACTIVITY_TYPE_CREATE_WALLET";
    ActivityType["ActivityTypeRemoveWallet"] = "ACTIVITY_TYPE_REMOVE_WALLET";
    ActivityType["ActivityTypeUpdateWallet"] = "ACTIVITY_TYPE_UPDATE_WALLET";
    ActivityType["ActivityTypeCastVote"] = "ACTIVITY_TYPE_CAST_VOTE";
    ActivityType["ActivityTypeEnableVoting"] = "ACTIVITY_TYPE_ENABLE_VOTING";
    ActivityType["ActivityTypeStake"] = "ACTIVITY_TYPE_STAKE";
    ActivityType["ActivityTypeUnstake"] = "ACTIVITY_TYPE_UNSTAKE";
    ActivityType["ActivityTypeChangeValidator"] = "ACTIVITY_TYPE_CHANGE_VALIDATOR";
    ActivityType["ActivityTypeRestake"] = "ACTIVITY_TYPE_RESTAKE";
    ActivityType["ActivityTypeAddressBook"] = "ACTIVITY_TYPE_ADDRESS_BOOK";
    ActivityType["ActivityTypeTeamMembers"] = "ACTIVITY_TYPE_TEAM_MEMBERS";
    ActivityType["ActivityTypeBilling"] = "ACTIVITY_TYPE_BILLING";
    ActivityType["ActivityTypeSecurity"] = "ACTIVITY_TYPE_SECURITY";
    ActivityType["ActivityTypeApi"] = "ACTIVITY_TYPE_API";
    ActivityType["ActivityTypeSettings"] = "ACTIVITY_TYPE_SETTINGS";
    ActivityType["ActivityTypeSmartContract"] = "ACTIVITY_TYPE_SMART_CONTRACT";
    ActivityType["ActivityTypeAllocationIn"] = "ACTIVITY_TYPE_ALLOCATION_IN";
    ActivityType["ActivityTypeAllocationOut"] = "ACTIVITY_TYPE_ALLOCATION_OUT";
    ActivityType["ActivityTypeAllocationInReversal"] = "ACTIVITY_TYPE_ALLOCATION_IN_REVERSAL";
    ActivityType["ActivityTypeAllocationOutReversal"] = "ACTIVITY_TYPE_ALLOCATION_OUT_REVERSAL";
    ActivityType["ActivityTypeConversion"] = "ACTIVITY_TYPE_CONVERSION";
    ActivityType["ActivityTypePrincipalOut"] = "ACTIVITY_TYPE_PRINCIPAL_OUT";
    ActivityType["ActivityTypePrincipalIn"] = "ACTIVITY_TYPE_PRINCIPAL_IN";
    ActivityType["ActivityTypeCollateralOut"] = "ACTIVITY_TYPE_COLLATERAL_OUT";
    ActivityType["ActivityTypeCollateralIn"] = "ACTIVITY_TYPE_COLLATERAL_IN";
    ActivityType["ActivityTypeInterestOut"] = "ACTIVITY_TYPE_INTEREST_OUT";
    ActivityType["ActivityTypeInterestIn"] = "ACTIVITY_TYPE_INTEREST_IN";
    ActivityType["ActivityTypeWeb3Message"] = "ACTIVITY_TYPE_WEB3_MESSAGE";
    ActivityType["ActivityTypeWeb3Transaction"] = "ACTIVITY_TYPE_WEB3_TRANSACTION";
    ActivityType["ActivityTypeWeb3DeviceRecovery"] = "ACTIVITY_TYPE_WEB3_DEVICE_RECOVERY";
    ActivityType["ActivityTypeWeb3RecreateBackup"] = "ACTIVITY_TYPE_WEB3_RECREATE_BACKUP";
    ActivityType["ActivityTypeWeb3Onboarding"] = "ACTIVITY_TYPE_WEB3_ONBOARDING";
})(ActivityType || (exports.ActivityType = ActivityType = {}));
