// @coinbase-sample/prime-sdk-ts/dist/model/enums/TransactionType.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * This file is generated by Openapi Generator https://github.com/openapitools/openapi-generator
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *  Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionType = void 0;
/**
 * - TRANSACTION_TYPE_UNKNOWN: An unknown transaction type  - DEPOSIT: A fiat or crypto deposit  - WITHDRAWAL: A fiat or crypto withdrawal  - INTERNAL_DEPOSIT: An internal fiat or crypto deposit  - INTERNAL_WITHDRAWAL: An internal fiat or crypto withdrawal  - SWEEP_DEPOSIT: Internal automated deposit to a cold address from a restored address  - SWEEP_WITHDRAWAL: Internal automated withdrawal from a restored address to a cold address  - PROXY_DEPOSIT: On-chain deposit of funds into proxy contract from cold address  - PROXY_WITHDRAWAL: On-chain withdrawal of funds from proxy contract to cold address  - BILLING_WITHDRAWAL: Coinbase Prime automated invoice settlement payment  - REWARD: Reward payment to an associated address for a staked asset  - COINBASE_REFUND: Coinbase Prime refund for the leftover amount for a CPFP (child pays for parent) transaction  - TRANSACTION_TYPE_OTHER: An OTHER type of transaction  - WITHDRAWAL_ADJUSTMENT: A manual adjustment withdrawal transaction  - DEPOSIT_ADJUSTMENT: A manual adjustment deposit transaction  - KEY_REGISTRATION: An on-chain registration for an address  - DELEGATION: An on-chain delegation transaction  - UNDELEGATION: An on-chain undelegation transaction  - RESTAKE: On-chain restaking transaction  - COMPLETE_UNBONDING: On-chain unbonding event transaction  - WITHDRAW_UNBONDED: On-chain event indicating unbonding period is over  - STAKE_ACCOUNT_CREATE: On-chain transaction to begin staking from an address  - CHANGE_VALIDATOR: On-chain transaction alter validator  - STAKE: On-chain transaction to begin staking in Cryptocurrency network  - UNSTAKE: On-chain transaction to stop staking in Cryptocurrency network  - REMOVE_AUTHORIZED_PARTY: On-chain transaction to remove a party from a multi-signature wallet  - STAKE_AUTHORIZE_WITH_SEED: On-chain transaction to begin staking from a seed account  - SLASH: On-chain transaction indicating a slash event has occurred  - COINBASE_DEPOSIT: On-chain transaction deposit for the purpose of transaction operations  - CONVERSION: Internal conversion between two assets  - CLAIM_REWARDS: On-chain transaction to claim rewards from Vote Account  - VOTE_AUTHORIZE: On-chain transaction to transfer the reward claiming permission to other pubkey  - WEB3_TRANSACTION: On-chain transaction initiated with Prime Onchain Wallet Deprecated: Use ONCHAIN_TRANSACTION instead  - ONCHAIN_TRANSACTION: On-chain transaction initiated with Prime Onchain Wallet
 */
var TransactionType;
(function (TransactionType) {
    TransactionType["Deposit"] = "DEPOSIT";
    TransactionType["Withdrawal"] = "WITHDRAWAL";
    TransactionType["InternalDeposit"] = "INTERNAL_DEPOSIT";
    TransactionType["InternalWithdrawal"] = "INTERNAL_WITHDRAWAL";
    TransactionType["SweepDeposit"] = "SWEEP_DEPOSIT";
    TransactionType["SweepWithdrawal"] = "SWEEP_WITHDRAWAL";
    TransactionType["ProxyDeposit"] = "PROXY_DEPOSIT";
    TransactionType["ProxyWithdrawal"] = "PROXY_WITHDRAWAL";
    TransactionType["BillingWithdrawal"] = "BILLING_WITHDRAWAL";
    TransactionType["Reward"] = "REWARD";
    TransactionType["CoinbaseRefund"] = "COINBASE_REFUND";
    TransactionType["TransactionTypeOther"] = "TRANSACTION_TYPE_OTHER";
    TransactionType["WithdrawalAdjustment"] = "WITHDRAWAL_ADJUSTMENT";
    TransactionType["DepositAdjustment"] = "DEPOSIT_ADJUSTMENT";
    TransactionType["KeyRegistration"] = "KEY_REGISTRATION";
    TransactionType["Delegation"] = "DELEGATION";
    TransactionType["Undelegation"] = "UNDELEGATION";
    TransactionType["Restake"] = "RESTAKE";
    TransactionType["CompleteUnbonding"] = "COMPLETE_UNBONDING";
    TransactionType["WithdrawUnbonded"] = "WITHDRAW_UNBONDED";
    TransactionType["StakeAccountCreate"] = "STAKE_ACCOUNT_CREATE";
    TransactionType["ChangeValidator"] = "CHANGE_VALIDATOR";
    TransactionType["Stake"] = "STAKE";
    TransactionType["Unstake"] = "UNSTAKE";
    TransactionType["RemoveAuthorizedParty"] = "REMOVE_AUTHORIZED_PARTY";
    TransactionType["StakeAuthorizeWithSeed"] = "STAKE_AUTHORIZE_WITH_SEED";
    TransactionType["Slash"] = "SLASH";
    TransactionType["CoinbaseDeposit"] = "COINBASE_DEPOSIT";
    TransactionType["Conversion"] = "CONVERSION";
    TransactionType["ClaimRewards"] = "CLAIM_REWARDS";
    TransactionType["VoteAuthorize"] = "VOTE_AUTHORIZE";
    TransactionType["Web3Transaction"] = "WEB3_TRANSACTION";
    TransactionType["OnchainTransaction"] = "ONCHAIN_TRANSACTION";
})(TransactionType || (exports.TransactionType = TransactionType = {}));
