// @coinbase-sample/prime-sdk-ts/dist/model/enums/UserRole.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * This file is generated by Openapi Generator https://github.com/openapitools/openapi-generator
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *  Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserRole = void 0;
/**
 * - USER_ROLE_UNKNOWN: nil value  - AUDITOR: An auditor  - SIGNATORY: A signatory  - ADMIN: An admin  - INITIATOR: An initiator  - REVIEWER: A reviewer  - TRADER: A trader  - FULL_TRADER: A trader with full permissions  - TEAM_MANAGER: A team manager  - APPROVER: An approver
 */
var UserRole;
(function (UserRole) {
    UserRole["Auditor"] = "AUDITOR";
    UserRole["Signatory"] = "SIGNATORY";
    UserRole["Admin"] = "ADMIN";
    UserRole["Initiator"] = "INITIATOR";
    UserRole["Reviewer"] = "REVIEWER";
    UserRole["Trader"] = "TRADER";
    UserRole["FullTrader"] = "FULL_TRADER";
    UserRole["TeamManager"] = "TEAM_MANAGER";
    UserRole["Approver"] = "APPROVER";
})(UserRole || (exports.UserRole = UserRole = {}));
