// @coinbase-sample/prime-sdk-ts/dist/model/enums/WalletDepositInstructionType.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * This file is generated by Openapi Generator https://github.com/openapitools/openapi-generator
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *  Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletDepositInstructionType = void 0;
/**
 * - UNKNOWN_WALLET_DEPOSIT_TYPE: nil value  - CRYPTO: A cryptocurrency deposit  - WIRE: A wire deposit  - SEN: DEPRECATED. A Silvergate Exchange Network deposit  - SWIFT: A SWIFT deposit  - SEPA: A SEPA deposit (Single Euro Payments Area)
 */
var WalletDepositInstructionType;
(function (WalletDepositInstructionType) {
    WalletDepositInstructionType["Crypto"] = "CRYPTO";
    WalletDepositInstructionType["Wire"] = "WIRE";
    WalletDepositInstructionType["Sen"] = "SEN";
    WalletDepositInstructionType["Swift"] = "SWIFT";
    WalletDepositInstructionType["Sepa"] = "SEPA";
})(WalletDepositInstructionType || (exports.WalletDepositInstructionType = WalletDepositInstructionType = {}));
