// @coinbase-sample/prime-sdk-ts/dist/model/enums/XMControlStatus.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * This file is generated by Openapi Generator https://github.com/openapitools/openapi-generator
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *  Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XMControlStatus = void 0;
/**
 * - TRADES_AND_WITHDRAWALS: Allowed to trade and withdraw. See XM Margin Methodology for full description of when trading and withdrawals are enabled or disabled.  - TRADES_ONLY: Allowed to trade but not withdraw. See XM Margin Methodology for full description of when trading and withdrawals are enabled or disabled.  - SESSION_LOCKED: Not allowed to trade or withdraw. See XM Margin Methodology for full description of when trading and withdrawals are enabled or disabled.
 */
var XMControlStatus;
(function (XMControlStatus) {
    XMControlStatus["XmControlStatusUnspecified"] = "XM_CONTROL_STATUS_UNSPECIFIED";
    XMControlStatus["TradesAndWithdrawals"] = "TRADES_AND_WITHDRAWALS";
    XMControlStatus["TradesOnly"] = "TRADES_ONLY";
    XMControlStatus["SessionLocked"] = "SESSION_LOCKED";
})(XMControlStatus || (exports.XMControlStatus = XMControlStatus = {}));
