// @coinbase-sample/prime-sdk-ts/dist/model/enums/XMEntityCallStatus.js
"use strict";
/**
 * Copyright 2025-present Coinbase Global, Inc.
 *
 * This file is generated by Openapi Generator https://github.com/openapitools/openapi-generator
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *  Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XMEntityCallStatus = void 0;
/**
 * - ENTITY_NO_CALL: There are no margin calls  - ENTITY_OPEN_STANDARD_CALL: There is a standard margin call. There are no urgent margin calls, and no expired margin calls.  - ENTITY_OPEN_URGENT_CALL: There is an urgent margin call. It is possible that there could also be a standard margin call, but there are no expired margin calls.  - ENTITY_AGED_CALL: At least one open margin call is aged (standard or urgent). This may trigger the SESSION_LOCKED control status.
 */
var XMEntityCallStatus;
(function (XMEntityCallStatus) {
    XMEntityCallStatus["XmEntityCallStatusUnspecified"] = "XM_ENTITY_CALL_STATUS_UNSPECIFIED";
    XMEntityCallStatus["EntityNoCall"] = "ENTITY_NO_CALL";
    XMEntityCallStatus["EntityOpenStandardCall"] = "ENTITY_OPEN_STANDARD_CALL";
    XMEntityCallStatus["EntityOpenUrgentCall"] = "ENTITY_OPEN_URGENT_CALL";
    XMEntityCallStatus["EntityAgedCall"] = "ENTITY_AGED_CALL";
})(XMEntityCallStatus || (exports.XMEntityCallStatus = XMEntityCallStatus = {}));
