// @coinbase-sample/prime-sdk-ts/dist/paymentMethods/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentMethodsService = void 0;
class PaymentMethodsService {
    constructor(client) {
        this.client = client;
    }
    // TODO: rename me to remove Entity in a minor version change
    listEntityPaymentMethods(request, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.request({
                url: `entities/${request.entityId}/payment-methods`,
                callOptions: options,
            });
            return response.data;
        });
    }
    getPaymentMethod(request, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.request({
                url: `entities/${request.entityId}/payment-methods/${request.paymentMethodId}`,
                callOptions: options,
            });
            return response.data;
        });
    }
}
exports.PaymentMethodsService = PaymentMethodsService;
