// @coinbase-sample/core-ts/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_PAGINATION_MAX_ITEMS = exports.DEFAULT_PAGINATION_MAX_PAGES = exports.DEFAULT_PAGINATION_LIMIT = exports.DEFAULT_HTTP_TIMEOUT = exports.USER_AGENT = exports.VERSION = void 0;
/**
 * Copyright 2024-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
exports.VERSION = '0.2.2';
exports.USER_AGENT = `coinbase-core-ts/${exports.VERSION}`;
exports.DEFAULT_HTTP_TIMEOUT = 5000;
exports.DEFAULT_PAGINATION_LIMIT = 25;
exports.DEFAULT_PAGINATION_MAX_PAGES = 100;
exports.DEFAULT_PAGINATION_MAX_ITEMS = 1000;
