// @coinbase-sample/core-ts/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonUtility = exports.Method = exports.CoinbaseHttpClient = exports.CoinbaseClientException = exports.CoinbaseError = exports.CoinbaseClient = void 0;
/**
 * Copyright 2024-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var client_1 = require("./client");
Object.defineProperty(exports, "CoinbaseClient", { enumerable: true, get: function () { return client_1.CoinbaseClient; } });
var error_1 = require("./error");
Object.defineProperty(exports, "CoinbaseError", { enumerable: true, get: function () { return error_1.CoinbaseError; } });
Object.defineProperty(exports, "CoinbaseClientException", { enumerable: true, get: function () { return error_1.CoinbaseClientException; } });
var httpClient_1 = require("./http/httpClient");
Object.defineProperty(exports, "CoinbaseHttpClient", { enumerable: true, get: function () { return httpClient_1.CoinbaseHttpClient; } });
var options_1 = require("./http/options");
Object.defineProperty(exports, "Method", { enumerable: true, get: function () { return options_1.Method; } });
var jsonUtility_1 = require("./serialization/jsonUtility");
Object.defineProperty(exports, "JsonUtility", { enumerable: true, get: function () { return jsonUtility_1.JsonUtility; } });
