// @coinbase-sample/prime-sdk-ts/dist/products/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductsService = void 0;
const paginatedResponse_1 = require("../shared/paginatedResponse");
class ProductsService {
    constructor(client) {
        this.client = client;
    }
    listProducts(request, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const paginationParams = (0, paginatedResponse_1.getQueryParams)(this.client, request);
            const { limit, cursor, sortDirection, portfolioId } = request, queryParams = __rest(request, ["limit", "cursor", "sortDirection", "portfolioId"]);
            const finalQueryParams = Object.assign(Object.assign({}, paginationParams), queryParams);
            const response = yield this.client.request({
                url: `portfolios/${portfolioId}/products`,
                queryParams: finalQueryParams,
                callOptions: options,
            });
            const paginationOptions = (0, paginatedResponse_1.getDefaultPaginationOptions)(this.client, options);
            return (0, paginatedResponse_1.createPaginatedResponse)(response.data, this.listProducts.bind(this), request, paginatedResponse_1.ResponseExtractors.products, paginationOptions);
        });
    }
    listProductCandles(request, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { portfolioId, productId, startTime, endTime, granularity } = request;
            const response = yield this.client.request({
                url: `portfolios/${portfolioId}/candles`,
                queryParams: {
                    product_id: productId,
                    start_time: startTime,
                    end_time: endTime,
                    granularity: granularity,
                },
                callOptions: options,
            });
            return response.data;
        });
    }
}
exports.ProductsService = ProductsService;
