// @coinbase-sample/core-ts/dist/http/coinbaseHttpRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinbaseHttpRequest = void 0;
/**
 * Copyright 2024-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const axios_1 = require("axios");
const constants_1 = require("../constants");
class CoinbaseHttpRequest {
    credentials;
    requestOptions;
    method;
    baseURL;
    url;
    data;
    headers;
    signal;
    callOptions;
    fullUrl;
    params;
    constructor(method, apiBasePath, requestPath = '', credentials, queryParams, bodyParams, callOptions) {
        this.credentials = credentials;
        const queryString = this.buildQueryString(queryParams);
        this.fullUrl = `${apiBasePath}${requestPath}${queryString}`;
        this.method = method;
        this.baseURL = apiBasePath;
        this.url = requestPath;
        this.callOptions = callOptions;
        this.data = bodyParams;
        this.params = this.sanitizeParams(this.buildURLSearchParams(queryParams));
        const headers = this.addAuthHeader();
        this.requestOptions = {
            method,
            headers,
            url: requestPath,
            data: bodyParams,
            signal: callOptions?.signal,
            timeout: callOptions?.timeout || constants_1.DEFAULT_HTTP_TIMEOUT,
        };
        this.headers = headers;
    }
    addAuthHeader() {
        const headers = new axios_1.AxiosHeaders();
        // sdk library responsibility to maintain public vs non-public endpoints
        if (this.credentials !== undefined) {
            const authHeaders = this.credentials.generateAuthHeaders(this.method, this.fullUrl, JSON.stringify(this.data) || '');
            headers.set(authHeaders);
        }
        return headers;
    }
    addHeader(key, value) {
        if (!(this.requestOptions.headers instanceof axios_1.AxiosHeaders)) {
            this.requestOptions.headers = new axios_1.AxiosHeaders();
        }
        this.requestOptions.headers.append(key, value);
        this.headers = this.requestOptions.headers;
    }
    filterParams(data) {
        const filteredParams = {};
        for (const key in data) {
            if (data[key] !== undefined) {
                filteredParams[key] = data[key];
            }
        }
        return filteredParams;
    }
    isValidValue(value) {
        return (value !== undefined &&
            value !== null &&
            value !== '' &&
            value !== 'null' &&
            value !== 'undefined');
    }
    buildURLSearchParams(queryParams) {
        const params = new URLSearchParams();
        if (!queryParams) {
            return params;
        }
        // Manually handle each parameter to support arrays
        for (const [key, value] of Object.entries(queryParams)) {
            if (Array.isArray(value)) {
                // Add each array element as a separate parameter with the same key
                value.forEach((item) => {
                    if (this.isValidValue(item)) {
                        params.append(key, String(item));
                    }
                });
            }
            else if (this.isValidValue(value)) {
                // Handle string values that might be comma-separated
                const stringValue = String(value);
                if (stringValue.includes(',')) {
                    // Split comma-separated values and add each as a separate parameter
                    stringValue.split(',').forEach((item) => {
                        const trimmedItem = item.trim();
                        if (trimmedItem) {
                            params.append(key, trimmedItem);
                        }
                    });
                }
                else {
                    params.append(key, stringValue);
                }
            }
        }
        return params;
    }
    buildQueryString(queryParams) {
        if (!queryParams || Object.keys(queryParams).length === 0) {
            return '';
        }
        const params = this.buildURLSearchParams(queryParams);
        const sanitizedParams = this.sanitizeParams(params);
        return `?${sanitizedParams.toString()}`;
    }
    sanitizeParams(params) {
        const emptyParams = [];
        params.forEach((value, key) => {
            if (!this.isValidValue(value)) {
                emptyParams.push(key);
            }
        });
        emptyParams.forEach((key) => {
            params.delete(key);
        });
        return params;
    }
}
exports.CoinbaseHttpRequest = CoinbaseHttpRequest;
