// @coinbase-sample/core-ts/dist/http/httpClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinbaseHttpClient = void 0;
/**
 * Copyright 2024-present Coinbase Global, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const axios_1 = __importDefault(require("axios"));
const axios_retry_1 = __importDefault(require("axios-retry"));
const coinbaseHttpRequest_1 = require("./coinbaseHttpRequest");
const options_1 = require("./options");
const error_1 = require("../error");
const constants_1 = require("../constants");
class CoinbaseHttpClient {
    credentials;
    httpClient;
    apiBasePath;
    userAgent;
    httpOptions;
    addedHeaders = {};
    addedRequestTransformers = [];
    addedResponseTransformers = [];
    constructor(apiBasePath, userAgent, credentials, options) {
        this.apiBasePath = apiBasePath;
        this.userAgent = userAgent;
        this.credentials = credentials;
        if (!options) {
            options = {
                defaultLimit: constants_1.DEFAULT_PAGINATION_LIMIT,
                maxPages: constants_1.DEFAULT_PAGINATION_MAX_PAGES,
                maxItems: constants_1.DEFAULT_PAGINATION_MAX_ITEMS,
            };
        }
        if (!options.defaultLimit)
            options.defaultLimit = constants_1.DEFAULT_PAGINATION_LIMIT;
        if (!options.maxPages)
            options.maxPages = constants_1.DEFAULT_PAGINATION_MAX_PAGES;
        if (!options.maxItems)
            options.maxItems = constants_1.DEFAULT_PAGINATION_MAX_ITEMS;
        this.httpOptions = options;
        this.httpClient = this._setupHttpClient(options);
    }
    _setupHttpClient(options) {
        const axiosClient = axios_1.default.create({
            baseURL: this.apiBasePath,
            headers: {
                'Content-Type': 'application/json',
                'User-Agent': this.userAgent,
            },
        });
        if (options) {
            if (options.retries) {
                (0, axios_retry_1.default)(axiosClient, { retries: options.retries });
            }
            else if (options.retryDelay) {
                (0, axios_retry_1.default)(axiosClient, {
                    retryDelay: axios_retry_1.default.linearDelay(options.retryDelay),
                });
            }
            else if (options.retryExponential) {
                (0, axios_retry_1.default)(axiosClient, {
                    retryDelay: axios_retry_1.default.exponentialDelay,
                });
            }
            else if (options.retryCustomFunction) {
                (0, axios_retry_1.default)(axiosClient, {
                    retryDelay: options.retryCustomFunction,
                });
            }
        }
        const transformRequest = options?.transformRequest
            ? Array.isArray(options.transformRequest)
                ? options.transformRequest
                : options.transformRequest
            : [];
        if (Array.isArray(transformRequest)) {
            transformRequest.forEach((transformer) => {
                axiosClient.interceptors.request.use(transformer, null);
                this.addedRequestTransformers.push(transformer);
            });
        }
        else if (typeof transformRequest === 'function') {
            axiosClient.interceptors.request.use(transformRequest, null);
            this.addedRequestTransformers.push(transformRequest);
        }
        const transformResponse = options?.transformResponse
            ? Array.isArray(options.transformResponse)
                ? options.transformResponse
                : options.transformResponse
            : [];
        if (Array.isArray(transformResponse)) {
            transformResponse.forEach((transformer) => {
                axiosClient.interceptors.response.use(transformer, null);
                this.addedResponseTransformers.push(transformer);
            });
        }
        else if (typeof transformResponse === 'function') {
            axiosClient.interceptors.response.use(transformResponse, null);
            this.addedResponseTransformers.push(transformResponse);
        }
        return axiosClient;
    }
    async sendRequest(options) {
        const { url, queryParams, bodyParams } = options;
        const requestMethod = options.method || options_1.Method.GET;
        const cbRequest = new coinbaseHttpRequest_1.CoinbaseHttpRequest(requestMethod, this.apiBasePath, url, this.credentials, queryParams, bodyParams, options.callOptions);
        let client = this.httpClient;
        if (options.callOptions) {
            //Does this need a custom transformer?
            const combinedOptions = {
                ...this.httpOptions,
                ...options.callOptions,
            };
            const callSpecificClient = this._setupHttpClient(combinedOptions);
            Object.entries(this.addedHeaders).forEach(([key, value]) => {
                callSpecificClient.defaults.headers[key] = value;
            });
            // Apply any transformers that were added to the main client
            this.addedResponseTransformers.forEach((transformer) => {
                callSpecificClient.interceptors.response.use(transformer, null);
            });
            this.addedRequestTransformers.forEach((transformer) => {
                callSpecificClient.interceptors.request.use(transformer, null);
            });
            client = callSpecificClient;
        }
        try {
            const response = await client.request(cbRequest);
            if (response?.headers && typeof response.headers.toJSON === 'function') {
                response.headers = response.headers.toJSON();
            }
            return response;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                (0, error_1.handleException)(error?.response, error.response?.data, error.message);
            }
            throw error;
        }
    }
    addHeader(key, value) {
        this.httpClient.defaults.headers[key] = value;
        this.addedHeaders[key] = value;
    }
    addTransformRequest(func) {
        this.addedRequestTransformers.push(func);
        this.httpClient.interceptors.request.use(func, null);
    }
    addTransformResponse(func) {
        this.addedResponseTransformers.push(func);
        this.httpClient.interceptors.response.use(func, null);
    }
    getDefaultPaginationLimit() {
        return this.httpOptions.defaultLimit;
    }
    getMaxPages() {
        return this.httpOptions.maxPages;
    }
    getMaxItems() {
        return this.httpOptions.maxItems;
    }
}
exports.CoinbaseHttpClient = CoinbaseHttpClient;
